lib/src/maintenance.dart

```diff
@@ -101,43 +101,135 @@ class Maintenance extends Object with _$MaintenanceSerializerMixin {
 
-    if (age > _twoYears) {
-      return 0.0;
-    }
-
-    var score = 1.0;
-
-    // adjust score to the age
-    if (age > _year) {
+    if (age != null) {
       final daysLeft = (_twoYears - age).inDays;
       final p = daysLeft / 365;
-      score *= max(0.0, min(1.0, p));
+      final score = max(0.0, min(1.0, p));
+
+      if (score == 0.0) {
+        explanations.add(ExplainScore._vintage);
+        return explanations;
+      } else if (score < 1.0) {
+        explanations.add(ExplainScore._outdate(score));
+      } else {
+        explanations.add(ExplainScore._updated);
+      }
     }
 
-    // missing files
-    if (missingChangelog) score *= 0.80;
-    if (missingReadme) score *= 0.95;
+    explanations.add(ExplainScore._readme(missingReadme));
+    explanations.add(ExplainScore._changelog(missingChangelog));
 
     // Pre-v1
     if (isExperimentalVersion) {
-      score *= 0.98;
+      explanations.add(new ExplainScore._(ExplainType.experimental, 0.98));
     }
 
     // Not a "gold" release
     if (isPreReleaseVersion) {
-      score *= 0.95;
+      explanations.add(new ExplainScore._(ExplainType.preRelease, 0.95));
+    }
+
+    // Stable!
+    if (!(isPreReleaseVersion || isExperimentalVersion)) {
+      explanations.add(new ExplainScore._(ExplainType.stable, 1.00));
     }
 
     // Bulk penalties. A few of these overlap with the penalties above, but the
     // difference is negligible, and not worth to compensate it.
-    score *= pow(0.80, errorCount);
-    score *= pow(0.99, warningCount);
-    score *= pow(0.999, hintCount);
+    explanations.add(new ExplainScore._(ExplainType.errors, 0.8, errorCount));
+    explanations
+        .add(new ExplainScore._(ExplainType.warnings, 0.99, warningCount));
+    explanations.add(new ExplainScore._(ExplainType.hints, 0.999, hintCount));
+
+    return explanations;
+  }
+
+  double getMaintenanceScore({Duration age}) => explainScore(age: age)
+      .fold(1.0, (previous, score) => previous * score.scale);
+}
+
+enum ExplainType {
+  /// More than 2 years since last update.
+  vintage,
+
+  /// More than 1 year since last update.
+  outdated,
+
+  /// Less than a year since the last update.
+  updated,
+
+  /// Package has a changelog.
+  hasChangelog,
+
+  /// Package is missing a changelog
+  missingChangelog,
 
-    return score;
+  /// Package has a readme.
+  hasReadme,
+
+  /// Package is missing a readme
+  missingReadme,
+
+  /// whether version is `0.*`
+  experimental,
+
+  /// Whether version is flagged `-beta`, `-alpha`, etc.
+  preRelease,
+
+  /// This is a stable release.
+  stable,
+
+  errors,
+
+  warnings,
+
+  hints,
+}
+
+final _explainTypeStrings = new Map<ExplainType, String>.fromIterable(
+    ExplainType.values,
+    key: (e) => e as ExplainType,
+    value: (e) => e.toString().split('.').last);
+
+class ExplainScore {
+  final ExplainType explanation;
+  final double simpleScale;
+  final int exponent;
+  final double scale;
```

Comment:
why not just a single score and an explanation?

Reply:
I want the "math" behind these to be 100% transparent. See toString. I want to put the math on the analysis page so folks understand how we get the numbers.

Reply:
I still think this is a bit more complicated than it needs to be. We could create a text log of what happens, alongside the result of the calculation, without the need of a new structure. It will be a human who will read it, no need to return a parsed structure or a pseudocode here.

We also have a `Suggestion` class which will be a duplicate for most of these: if there is a penalty, we need to explain it and suggest something to be fixed. Maybe this log could be part of the description in a suggestion?